libcudacxx/include/cuda/std/__atomic/functions/cuda_ptx_derived.h

```diff
@@ -0,0 +1,191 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of libcu++, the C++ Standard Library for your entire system,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef __LIBCUDACXX___ATOMIC_FUNCTIONS_DERIVED_H
+#define __LIBCUDACXX___ATOMIC_FUNCTIONS_DERIVED_H
+
+#include <cuda/std/detail/__config>
+
+#include <cuda/std/__atomic/functions/cuda_ptx_generated.h>
+#include <cuda/std/cstdint>
+
+_LIBCUDACXX_BEGIN_NAMESPACE_STD
+
+template <typename _Tp, typename _Sco, __enable_if_t<sizeof(_Tp) <= 2, int> = 0>
+bool _CCCL_DEVICE __atomic_compare_exchange_cuda(
+  _Tp volatile* __ptr, _Tp* __expected, const _Tp __desired, bool, int __success_memorder, int __failure_memorder, _Sco)
+{
+  auto const __aligned = (uint32_t*) ((intptr_t) __ptr & ~(sizeof(uint32_t) - 1));
+  auto const __offset  = uint32_t((intptr_t) __ptr & (sizeof(uint32_t) - 1)) * 8;
+  auto const __mask    = ((1 << sizeof(_Tp) * 8) - 1) << __offset;
+
+  uint32_t __old = *__expected << __offset;
+  uint32_t __old_value;
+  while (1)
+  {
+    __old_value = (__old & __mask) >> __offset;
+    if (__old_value != *__expected)
+    {
+      break;
+    }
+    uint32_t const __attempt = (__old & ~__mask) | (*__desired << __offset);
+    if (__atomic_compare_exchange_cuda(
+          __aligned, &__old, &__attempt, true, __success_memorder, __failure_memorder, _Sco{}))
+    {
+      return true;
+    }
+  }
+  *__expected = __old_value;
+  return false;
+}
+
+template <typename _Tp, typename _Sco, __enable_if_t<sizeof(_Tp) <= 2, int> = 0>
+void _CCCL_DEVICE __atomic_exchange_cuda(_Tp volatile* __ptr, _Tp* __val, _Tp* __ret, int __memorder, _Sco)
+{
+  _Tp __expected = __atomic_load_n_cuda(__ptr, __ATOMIC_RELAXED, _Sco{});
+  while (!__atomic_compare_exchange_cuda(__ptr, &__expected, __val, true, __memorder, __memorder, _Sco{}))
+    ;
```

Comment:
Please add `{}`, Also should this have a backoff?

Reply:
Atomic exchange shouldn't have a backoff. We aren't polling in this case, we're fighting another thread operating on the memory location. I believe it's a point of contention that we should nanosleep in order to prevent contention, but that would require testing and algorithms that rely on that behavior should implement some form of coalescing.